File: CommonUtils/src/com/rarnu/utils/FileUtils.java
package com.rarnu.utils;

import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;

import android.content.Context;
import android.os.Handler;
import android.os.Message;

public class FileUtils {

	public static final int WHAT_COPY_START = 1;
	public static final int WHAT_COPY_PROGRESS = 2;
	public static final int WHAT_COPY_FINISH = 3;

	public static boolean mkdir(String path) {
		boolean ret = false;
		File myDir = new File(path);
		if (!myDir.exists()) {
			ret = myDir.mkdirs();
		}
		return ret;
	}

	public static void rewriteFile(File file, String text, String encoding)
			throws IOException {
		writeFileByStream(file, text, false, encoding);
	}

	public static void rewriteFile(File file, String text) throws IOException {
		writeFileByWriter(file, text, false);
	}

	public static void rewriteFile(String path, String text) throws IOException {
		File myFile = new File(path);
		rewriteFile(myFile, text);
	}

	public static void rewriteFile(String path, String text, String encoding)
			throws IOException {
		File myFile = new File(path);
		rewriteFile(myFile, text, encoding);
	}

	public static void appendFile(File file, String text) throws IOException {
		writeFileByWriter(file, text, true);
	}

	public static void appendFile(File file, String text, String encoding)
			throws IOException {
		writeFileByStream(file, text, true, encoding);
	}

	private static void writeFileByWriter(File file, String text, boolean append)
			throws IOException {
		FileWriter myFileWriter = new FileWriter(file, append);
		if (append) {
			myFileWriter.append(text);
		} else {
			myFileWriter.write(text);
		}
		myFileWriter.close();
	}

	private static void writeFileByStream(File file, String text,
			boolean append, String encoding) throws IOException {
		FileOutputStream fos = new FileOutputStream(file, append);
		OutputStreamWriter myWriter = new OutputStreamWriter(fos, encoding);
		myWriter.write(text);
		myWriter.close();
		fos.close();
	}

	public static void appendFile(String path, String text) throws IOException {
		File myFile = new File(path);
		appendFile(myFile, text);
	}

	public static void appendFile(String path, String text, String encoding)
			throws IOException {
		File myFile = new File(path);
		appendFile(myFile, text, encoding);
	}

	public static boolean deleteFile(String path) {
		File myFile = new File(path);
		return myFile.delete();
	}

	public static boolean deleteDir(String path) {
		deleteSubFiles(path);
		File myDir = new File(path);
		return myDir.delete();
	}

	public static void deleteSubFiles(String path) {
		File myFile = new File(path);
		if (!myFile.exists()) {
			return;
		}
		if (!myFile.isDirectory()) {
			return;
		}
		String[] tempList = myFile.list();
		File temp = null;
		for (int i = 0; i < tempList.length; i++) {
			if (path.endsWith(File.separator)) {
				temp = new File(path + tempList[i]);
			} else {
				temp = new File(path + File.separator + tempList[i]);
			}
			if (temp.isFile()) {
				temp.delete();
			}
			if (temp.isDirectory()) {
				deleteSubFiles(path + File.separator + tempList[i]);
				deleteDir(path + File.separator + tempList[i]);
			}
		}
	}

	public static void copyFile(String source, String dest, Handler hProgress)
			throws IOException {

		File oldFile = new File(source);
		if (oldFile.exists()) {
			InputStream is = new FileInputStream(source);
			FileOutputStream fs = new FileOutputStream(dest);

			int size = is.available();
			if (hProgress != null) {
				Message msg = new Message();
				msg.what = WHAT_COPY_START;
				msg.arg1 = 0;
				msg.arg2 = size;
				hProgress.sendMessage(msg);
			}
			int count = 0;
			int n = 0;

			byte[] buffer = new byte[1444];
			while ((n = is.read(buffer)) != -1) {
				fs.write(buffer, 0, n);
				count += n;
				if (hProgress != null) {
					Message msg = new Message();
					msg.what = WHAT_COPY_PROGRESS;
					msg.arg1 = count;
					msg.arg2 = size;
					hProgress.sendMessage(msg);
				}
			}
			is.close();
			fs.close();
			if (hProgress != null) {
				Message msg = new Message();
				msg.what = WHAT_COPY_FINISH;
				msg.arg1 = 0;
				msg.arg2 = 0;
				hProgress.sendMessage(msg);
			}
		}
	}

	public static void copyFolder(String source, String dest)
			throws IOException {
		new File(dest).mkdirs();
		File a = new File(source);
		String[] file = a.list();
		File temp = null;
		for (int i = 0; i < file.length; i++) {
			if (source.endsWith(File.separator)) {
				temp = new File(source + file[i]);
			} else {
				temp = new File(source + File.separator + file[i]);
			}

			if (temp.isFile()) {
				FileInputStream input = new FileInputStream(temp);
				FileOutputStream output = new FileOutputStream(dest
						+ File.separator + (temp.getName()).toString());
				byte[] b = new byte[1024 * 5];
				int len;
				while ((len = input.read(b)) != -1) {
					output.write(b, 0, len);
				}
				output.flush();
				output.close();
				input.close();
			}
			if (temp.isDirectory()) {
				copyFolder(source + File.separator + file[i], dest
						+ File.separator + file[i]);
			}
		}

	}

	public static void moveFile(String source, String dest, Handler hProgress)
			throws IOException {
		copyFile(source, dest, hProgress);
		deleteFile(source);

	}

	public static void moveFolder(String source, String dest)
			throws IOException {
		copyFolder(source, dest);
		deleteDir(source);

	}

	public static List<String> readFile(File file) throws IOException {
		FileReader myFileReader = new FileReader(file);
		BufferedReader myBufferedReader = new BufferedReader(myFileReader);
		String line;
		List<String> fileText = new ArrayList<String>();
		while ((line = myBufferedReader.readLine()) != null) {
			fileText.add(line);
		}
		myBufferedReader.close();
		myFileReader.close();
		return fileText;
	}

	public static String toString(List<String> list) {
		String ret = "";
		for (String s : list) {
			ret += s + "\n";
		}
		return ret;
	}

	public static List<String> readFile(String path) throws IOException {
		File myFile = new File(path);
		return readFile(myFile);
	}

	public static String readAssetFile(Context context, String fileName)
			throws IOException {
		InputStream is = context.getAssets().open(fileName);
		byte[] bytes = new byte[1024];
		ByteArrayOutputStream arrayOutputStream = new ByteArrayOutputStream();
		while (is.read(bytes) != -1) {
			arrayOutputStream.write(bytes, 0, bytes.length);
		}
		is.close();
		arrayOutputStream.close();
		String text = new String(arrayOutputStream.toByteArray());
		return text.trim();
	}

	public static List<String> readAssertFileAsList(Context context,
			String fileName) throws IOException {
		InputStream is = context.getAssets().open(fileName);
		InputStreamReader myStreamReader = new InputStreamReader(is);
		BufferedReader myBufferedReader = new BufferedReader(myStreamReader);
		String line;
		List<String> fileText = new ArrayList<String>();
		while ((line = myBufferedReader.readLine()) != null) {
			fileText.add(line);
		}
		myBufferedReader.close();
		myStreamReader.close();
		return fileText;
	}

	public static boolean copyAssetFile(Context context, String fileName,
			String saveDir, Handler hProgress) {
		File fAsset = new File(saveDir);
		if (!fAsset.exists()) {
			fAsset.mkdirs();
		}
		try {
			byte[] buffer = new byte[8192];

			File dest = new File(saveDir + fileName);

			if (dest.exists()) {
				dest.delete();
			}

			InputStream is = context.getAssets().open(fileName);
			OutputStream fos = new BufferedOutputStream(new FileOutputStream(
					dest));

			int count = 0;
			int size = is.available();
			if (hProgress != null) {
				Message msg = new Message();
				msg.what = WHAT_COPY_START;
				msg.arg1 = 0;
				msg.arg2 = size;
				hProgress.sendMessage(msg);
			}
			int n;
			while ((n = is.read(buffer, 0, buffer.length)) != -1) {
				fos.write(buffer, 0, n);
				count += n;
				if (hProgress != null) {
					Message msg = new Message();
					msg.what = WHAT_COPY_PROGRESS;
					msg.arg1 = count;
					msg.arg2 = size;
					hProgress.sendMessage(msg);
				}
			}

			is.close();
			fos.close();
			if (hProgress != null) {
				Message msg = new Message();
				msg.what = WHAT_COPY_FINISH;
				msg.arg1 = 0;
				msg.arg2 = 0;
				hProgress.sendMessage(msg);
			}
			return true;
		} catch (Exception ex) {
			return false;
		}
	}

	public static String readInnerFile(Context context, String path)
			throws IOException {
		InputStream is = context.openFileInput(path);
		return readFileStream(is);
	}

	public static String readFileString(String path) throws IOException {
		InputStream is = new FileInputStream(path);
		return readFileStream(is);
	}

	public static String readFileString(String path, String encoding)
			throws IOException {
		String ret = readFileString(path);
		return new String(ret.getBytes(), encoding);
	}

	private static String readFileStream(InputStream is) throws IOException {
		byte[] bytes = new byte[1024];
		ByteArrayOutputStream arrayOutputStream = new ByteArrayOutputStream();
		while (is.read(bytes) != -1) {
			arrayOutputStream.write(bytes, 0, bytes.length);
		}
		is.close();
		arrayOutputStream.close();
		String text = new String(arrayOutputStream.toByteArray());
		return text;
	}

	public static List<?> loadListFromFile(String path) {
		List<?> list = null;
		try {
			FileInputStream freader = new FileInputStream(path);
			ObjectInputStream objectInputStream = new ObjectInputStream(freader);
			list = (List<?>) objectInputStream.readObject();
			objectInputStream.close();
		} catch (Exception e) {

		}
		return list;
	}

	public static void saveListToFile(List<?> list, String path) {
		try {
			FileOutputStream outStream = new FileOutputStream(path);
			ObjectOutputStream objectOutputStream = new ObjectOutputStream(
					outStream);
			objectOutputStream.writeObject(list);
			outStream.close();
		} catch (Exception e) {

		}
	}

	public static long getDirSize(String path) {
		return getDirSize(new File(path));
	}

	public static long getDirSize(File dir) {
		if (dir == null) {
			return 0;
		}
		if (!dir.isDirectory()) {
			return 0;
		}
		long dirSize = 0;
		File[] files = dir.listFiles();
		for (File file : files) {
			if (file.isFile()) {
				dirSize += file.length();
			} else if (file.isDirectory()) {
				dirSize += file.length();
				dirSize += getDirSize(file);
			}
		}
		return dirSize;
	}

}

File: MI8DoubleApp/src/com/rarnu/mi8/FileUtils.kt
package com.rarnu.mi8

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.os.Handler
import android.os.StatFs
import android.util.Log
import java.io.*
import java.nio.charset.Charset
import java.text.DecimalFormat

/**
 * Created by rarnu on 3/25/16.
 */
object FileUtils {

    val S_IRU = 400
    val S_IRG = 40
    val S_IRO = 4
    val S_IWU = 200
    val S_IWG = 20
    val S_IWO = 2
    val S_IXU = 100
    val S_IXG = 10
    val S_IXO = 1
    val S_IRWU = S_IRU or S_IWU
    val S_IRWG = S_IRG or S_IWG
    val S_IRWO = S_IRO or S_IWO
    val S_IRXU = S_IRU or S_IXU
    val S_IRXG = S_IRG or S_IXG
    val S_IRXO = S_IRO or S_IXO
    val S_IRWXU = S_IRU or S_IWU or S_IXU
    val S_IRWXG = S_IRG or S_IWG or S_IXG
    val S_IRWXO = S_IRO or S_IWO or S_IXO

    fun mkdir(path: String): Boolean {
        var ret = false
        val d = File(path)
        if (!d.exists()) {
            ret = d.mkdirs()
        }
        return ret
    }

    fun rewriteFile(file: File?, text: String, encoding: String) = writeFileByStream(file, text, false, encoding)

    fun rewriteFile(file: File?, text: String) = writeFileByWriter(file, text, false)

    fun rewriteFile(path: String, text: String) = rewriteFile(File(path), text)

    fun rewriteFile(path: String, text: String, encoding: String) = rewriteFile(File(path), text, encoding)

    fun appendFile(file: File?, text: String) = writeFileByWriter(file, text, true)

    fun appendFile(file: File?, text: String, encoding: String) = writeFileByStream(file, text, true, encoding)

    private fun writeFileByWriter(file: File?, text: String, append: Boolean) {
        val writer = FileWriter(file, append)
        if (append) {
            writer.append(text)
        } else {
            writer.write(text)
        }
        writer.close()
    }

    private fun writeFileByStream(file: File?, text: String, append: Boolean, encoding: String) {
        val fos = FileOutputStream(file, append)
        val writer = OutputStreamWriter(fos, encoding)
        writer.write(text)
        writer.close()
        fos.close()
    }

    fun appendFile(path: String, text: String) = appendFile(File(path), text)

    fun appendFile(path: String, text: String, encoding: String) = appendFile(File(path), text, encoding)

    fun deleteFile(path: String): Boolean = File(path).delete()

    fun deleteDir(path: String): Boolean {
        deleteSubFiles(path)
        return File(path).delete()
    }

    fun deleteSubFiles(path: String) {
        val f = File(path)
        if (!f.exists()) {
            return
        }
        if (!f.isDirectory) {
            return
        }
        val tempList = f.list()
        var temp: File?
        for (t in tempList) {
            temp = File(path, t)
            if (temp.isFile) {
                temp.delete()
            }
            if (temp.isDirectory) {
                deleteSubFiles(path + File.separator + t)
                temp.delete()
            }
        }
    }

    fun copyFile(source: String, dest: String, hProgress: Handler?) {
        val oldFile = File(source)
        if (oldFile.exists()) {
            val ins = FileInputStream(source)
            val fos = FileOutputStream(dest)
            val size = ins.available()
            MessageUtils.sendMessage(hProgress, Actions.WHAT_COPY_START, 0, size)
            var count = 0
            var n: Int
            val buffer = ByteArray(1024)
            while (true) {
                n = ins.read(buffer)
                if (n != -1) {
                    fos.write(buffer, 0, n)
                    count += n
                    MessageUtils.sendMessage(hProgress, Actions.WHAT_COPY_PROGRESS, count, size)
                } else {
                    break
                }
            }
            ins.close()
            fos.close()
            MessageUtils.sendMessage(hProgress, Actions.WHAT_COPY_FINISH)
        }
    }

    fun copyFolder(source: String, dest: String) {
        File(dest).mkdirs()
        val a = File(source)
        val file = a.list()
        var temp: File?
        for (f in file) {
            temp = File(source, f)
            if (temp.isFile) {
                val input = FileInputStream(temp)
                val output = FileOutputStream(dest + File.separator + temp.name)
                val b = ByteArray(1024)
                var len: Int
                while (true) {
                    len = input.read(b)
                    if (len != -1) {
                        output.write(b, 0, len)
                    } else {
                        break
                    }
                }
                output.flush()
                output.close()
                input.close()
            }
            if (temp.isDirectory) {
                copyFolder(source + File.separator + f, dest + File.separator + f)
            }
        }
    }

    fun moveFile(source: String, dest: String, hProgress: Handler?) {
        copyFile(source, dest, hProgress)
        deleteFile(source)
    }

    fun moveFolder(source: String, dest: String) {
        copyFolder(source, dest)
        deleteDir(source)
    }

    fun renameFile(source: String, dest: String) {
        val fSource = File(source)
        val fDest = File(dest)
        fSource.renameTo(fDest)
    }

    fun readFile(file: File?): MutableList<String>? {
        val reader = FileReader(file)
        val bufReader = BufferedReader(reader)
        var line: String?
        val fileText = arrayListOf<String>()
        while (true) {
            line = bufReader.readLine()
            if (line != null) {
                fileText.add(line)
            } else {
                break
            }
        }
        bufReader.close()
        reader.close()
        return fileText
    }

    fun fileList2string(list: MutableList<String>?): String? {
        var ret = ""
        for (s in list!!) {
            ret += "$s\n"
        }
        return ret
    }

    fun readFile(path: String): MutableList<String>? = readFile(File(path))

    fun readAssetFile(context: Context, fileName: String): String? {
        val ins = context.assets.open(fileName)
        val buffer = ByteArray(1024)
        val output = ByteArrayOutputStream()
        var len: Int
        while (true) {
            len = ins.read(buffer)
            if (len != -1) {
                output.write(buffer, 0, len)
            } else {
                break
            }
        }
        ins.close()
        output.close()
        val text = String(output.toByteArray())
        return text.trim()
    }

    fun readAssertFileAsList(context: Context, fileName: String): MutableList<String>?  {
        val ins = context.assets.open(fileName)
        val reader = InputStreamReader(ins)
        val bufReader = BufferedReader(reader)
        var line: String?
        var fileText = arrayListOf<String>()
        while (true) {
            line = bufReader.readLine()
            if (line != null) {
                fileText.add(line)
            } else {
                break
            }
        }
        ins.close()
        bufReader.close()
        reader.close()
        return fileText
    }

    fun copyAssetFile(context: Context, fileName: String, saveDir: String, hProgress: Handler?): Boolean {
        val fAsset = File(saveDir)
        if (!fAsset.exists()) {
            fAsset.mkdirs()
        }
        try {
            val buffer = ByteArray(1024)
            val dest = File(saveDir, fileName)
            if (dest.exists()) {
                dest.delete()
            }
            val ins = context.assets.open(fileName)
            val fos = BufferedOutputStream(FileOutputStream(dest))
            var count = 0
            val size = ins.available()
            MessageUtils.sendMessage(hProgress, Actions.WHAT_COPY_START, 0, size)
            var n: Int
            while (true) {
                n = ins.read(buffer, 0, buffer.size)
                if (n != -1) {
                    fos.write(buffer, 0, n)
                    count += n
                    MessageUtils.sendMessage(hProgress, Actions.WHAT_COPY_PROGRESS, count, size)
                } else {
                    break
                }
            }
            ins.close()
            fos.close()
            MessageUtils.sendMessage(hProgress, Actions.WHAT_COPY_FINISH)
            return true
        } catch (e: Exception) {
            return false
        }
    }

    fun readInnerFile(context: Context, path: String): String? = readFileStream(context.openFileInput(path))

    fun readFileString(path: String): String? = readFileStream(FileInputStream(path))

    fun readFileString(path: String, encoding: String): String? = String(readFileString(path)!!.toByteArray(), Charset.forName(encoding))

    private fun readFileStream(ins: InputStream?): String? {
        val buffer = ByteArray(1024)
        val output = ByteArrayOutputStream()
        var len: Int
        while (true) {
            len = ins!!.read(buffer)
            if (len != -1) {
                output.write(buffer, 0, len)
            } else {
                break
            }
        }
        ins.close()
        output.close()
        val text = String(output.toByteArray())
        return text.trim()
    }

    fun getDirSize(path: String): Long = getDirSize(File(path))

    fun getDirSize(dir: File?): Long {
        if (dir == null) {
            return 0
        }
        if (!dir.isDirectory) {
            return 0
        }
        var dirSize = 0L
        val files = dir.listFiles()
        for (file in files) {
            if (file.isFile) {
                dirSize += file.length()
            } else if (file.isDirectory) {
                dirSize += file.length()
                dirSize += getDirSize(file)
            }
        }
        return dirSize
    }


    fun getFileSize(path: String): Long = File(path).length()

    fun getReadableFileSize(size: Long): String {
        val units = arrayOf("B", "KB", "MB", "GB", "TB", "PB")
        var nSize = size * 1L * 1.0F
        val mod = 1024.0F
        var i = 0
        while (nSize >= mod) {
            nSize /= mod
            i++
        }
        val df = DecimalFormat(".00")
        return "${df.format(nSize)} ${units[i]}"
    }

    @Suppress("DEPRECATION")
    fun getTotalDataSzie(): Long {
        val path = Environment.getDataDirectory()
        val stat = StatFs(path.path)
        val blockSize: Long
        val blockCount: Long
        if (Build.VERSION.SDK_INT >= 21) {
            blockSize = stat.blockSizeLong
            blockCount = stat.blockCountLong
        } else {
            blockSize = stat.blockSize.toLong()
            blockCount = stat.blockCount.toLong()
        }
        return Math.abs(blockSize * blockCount)
    }

    @Suppress("DEPRECATION")
    fun getAvailableDataSize(): Long {
        val path = Environment.getDataDirectory()
        val stat = StatFs(path.path)
        val blockSize: Long
        val blocks: Long
        if (Build.VERSION.SDK_INT >= 21) {
            blockSize = stat.blockSizeLong
            blocks = stat.availableBlocksLong
        } else {
            blockSize = stat.blockSize.toLong()
            blocks = stat.availableBlocks.toLong()
        }
        return Math.abs(blocks * blockSize)
    }

    @Suppress("DEPRECATION")
    fun getTotalSdcardSize(): Long {
        val path = Environment.getExternalStorageDirectory()
        val stat = StatFs(path.path)
        val blockSize: Long
        val totalBlocks: Long
        if (Build.VERSION.SDK_INT >= 21) {
            blockSize = stat.blockSizeLong
            totalBlocks = stat.blockCountLong
        } else {
            blockSize = stat.blockSize.toLong()
            totalBlocks = stat.blockCount.toLong()
        }
        return Math.abs(blockSize * totalBlocks)
    }

    @Suppress("DEPRECATION")
    fun getAvailableSdcardSize(): Long {
        val path = Environment.getExternalStorageDirectory()
        val stat = StatFs(path.path)
        val blockSize: Long
        val availableBlocks: Long
        if (Build.VERSION.SDK_INT >= 21) {
            blockSize = stat.blockSizeLong
            availableBlocks = stat.availableBlocksLong
        } else {
            blockSize = stat.blockSize.toLong()
            availableBlocks = stat.availableBlocks.toLong()
        }
        return Math.abs(blockSize * availableBlocks)
    }

    fun getPackageSize(context: Context): Long {
        var size = 0L
        try {
            val info = context.packageManager.getApplicationInfo(context.packageName, 0)
            val dir = info.dataDir
            size = getDirSize(File(dir))
        } catch(e: Exception) {

        }
        return Math.abs(size)
    }

    fun loadListFromFile(path: String): MutableList<*>? = loadObjetFromFile(path) as MutableList<*>

    fun saveListToFile(list: MutableList<*>, path: String) = saveObjectToFile(list, path)

    fun loadObjetFromFile(path: String): Any? {
        var obj: Any? = null
        try {
            val input = FileInputStream(path)
            val ois = ObjectInputStream(input)
            obj = ois.readObject()
            ois.close()
            input.close()
        } catch (e: Exception) {

        }
        return obj
    }

    fun saveObjectToFile(obj: Any?, path: String) {
        try {
            val output = FileOutputStream(path)
            val oos = ObjectOutputStream(output)
            oos.writeObject(obj)
            oos.close()
            output.close()
        } catch (e: Exception) {

        }
    }

    fun setPermission(filePath: String, permission: Int) {
        try {
            var cls = Class.forName("android.os.FileUtils")
            var method = cls.getDeclaredMethod("setPermissions", String::class.java, Integer.TYPE, Integer.TYPE, Integer.TYPE)
            var ret = method.invoke(null, filePath, permission, -1, -1)
            Log.e("LOG", "setPermission => $ret")
        } catch(e: Exception) {
            Log.e("LOG", "setPermission => ${e.message}")
        }
    }

    fun sendScanCmd(context: Context, path: String) {
        context.sendBroadcast(Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.parse("file://$path")))
    }

}